<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<HTML>
 
<!-- Mirrored from www.py.cz/pipermail/python/2012-May/010679.html by HTTrack Website Copier/3.x [XR&CO'2014], Sun, 20 Mar 2016 21:17:31 GMT -->
<HEAD>
   <TITLE> [python] Unicode/ne-unicode problem
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:python%40py.cz?Subject=Re%3A%20%5Bpython%5D%20Unicode/ne-unicode%20problem&In-Reply-To=%3C0c5f01cd285a%2481601da0%246164a8c0%40IBMXP%3E">
   <META NAME="robots" CONTENT="index,nofollow">
   <style type="text/css">
       pre {
           white-space: pre-wrap;       /* css-2.1, curent FF, Opera, Safari */
           }
   </style>
   <META http-equiv="Content-Type" content="text/html; charset=iso-8859-2">
   <LINK REL="Previous"  HREF="http://www.py.cz/pipermail/python/2012-May/010678.html">
   <LINK REL="Next"  HREF="http://www.py.cz/pipermail/python/2012-May/010680.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>[python] Unicode/ne-unicode problem</H1>
    <B>Jaroslav Lukesh</B> 
    <A HREF="mailto:python%40py.cz?Subject=Re%3A%20%5Bpython%5D%20Unicode/ne-unicode%20problem&In-Reply-To=%3C0c5f01cd285a%2481601da0%246164a8c0%40IBMXP%3E"
       TITLE="[python] Unicode/ne-unicode problem">lukesh na seznam.cz
       </A><BR>
    <I>Støeda Kvìten  2 13:55:44 CEST 2012</I>
    <P><UL>
        <LI>Pøedchozí pøíspìvek: <A HREF="http://www.py.cz/pipermail/python/2012-May/010678.html">[python] Unicode/ne-unicode problem
</A></li>
        <LI>Následující pøíspìvek: <A HREF="http://www.py.cz/pipermail/python/2012-May/010680.html">[python] Unicode/ne-unicode problem
</A></li>
         <LI> <B>Zprávy tøídìné podle:</B> 
              <a href="http://www.py.cz/pipermail/python/2012-May/date.html#10679">[ data ]</a>
              <a href="http://www.py.cz/pipermail/python/2012-May/thread.html#10679">[ vlákna ]</a>
              <a href="http://www.py.cz/pipermail/python/2012-May/subject.html#10679">[ subjektu ]</a>
              <a href="http://www.py.cz/pipermail/python/2012-May/author.html#10679">[ autora ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>Myslím ¾e jsem to zkopíroval celé, ale mohlo mi nìco utéct. Jen¾e teï to mám 
z ulo¾ených poznámek v PSpadu, tak ten do toho taky kecal a teï je to úplnì 
jinak, ne¾ to co jsem v pondìlí kopíroval pøímo do promptu, ach jo.

No nic, musím to znova vygenerovat.

  ----- Pùvodní zpráva ----- 
  Od: Petr Messner
  Komu: Konference PyCZ
  Odesláno: 2. kvìtna 2012 13:33
  Pøedmìt: Re: [python] Unicode/ne-unicode problem


  Ono to mo¾ná nikam nevede. Bez posledních dvou znakù je to OK:


  &gt;&gt;&gt; print &quot;\xC3\xBD \xC3\xB7 \xC3\xAD&quot;.decode(&quot;utf8&quot;)
  ý ÷ í


  Ono za \xC3 musí z definice UTF-8 následovat bajt, který má nastavený 
nejvy¹¹í bit, a to obyèejná teèka není snad v ¾ádném kódování, rozhodnì v 
¾ádném zalo¾eném na ASCII jako je cp1250, latin1, latin2 atd.


  Prostì si myslím, ¾e tohle: 'Ã½ Ã· Ã­ Ã.' je ¹patnì, nìjak useknuté, nìco 
vynechané apod.


  Mù¾ete nám ten zdroj dat v UTF-8 nìjak popsat? Je to nìjaká databáze?


  PM




  Dne 2. kvìtna 2012 13:08 Jaroslav Lukesh &lt;<A HREF="http://www.py.cz/mailman/listinfo/python">lukesh na seznam.cz</A>&gt; napsal(a):

    Mezitím jsme v jiném threadu zjistili, ¾e jde o utf8 zapsané v cp1250, 
ale i tak mám tohle - Python 2.4.6:



          a='Ã½ Ã· Ã­ Ã.'
          repr(a)

    &quot;'\\xc3\\xbd \\xc3\\xb7 \\xc3\\xad \\xc3.'&quot;

          a.decode(&quot;cp1250&quot;)

    u'\u0102\u02dd \u0102\xb7 \u0102\xad \u0102.'

          print eval(a).decode(&quot;cp1250&quot;)

    Traceback (most recent call last):
     File &quot;&lt;stdin&gt;&quot;, line 1, in ?

     File &quot;&lt;string&gt;&quot;, line 1
      Ã½ Ã· Ã­ Ã.
      ^
    SyntaxError: invalid syntax




    No teï u¾ jenom to reverznout z cp1250 na utf8 a to pak u¾ správnì 
pøevést na cp1250. jen¾e necode/decode neumí vzít 2 parametry pro kódování, 
tak se to pokusím zøetìzit:


          a.decode('cp1250')

    u'\u0102\u02dd \u0102\xb7 \u0102\xad \u0102.'

          a.decode('cp1250').encode('utf8')

    '\xc4\x82\xcb\x9d \xc4\x82\xc2\xb7 \xc4\x82\xc2\xad \xc4\x82.'

    Sice jsem pokroèil, ale je to nìjak del¹í a del¹í.


    ----- Pùvodní zpráva ----- Od: &quot;Petr Pøikryl&quot; &lt;<A HREF="http://www.py.cz/mailman/listinfo/python">prikryl na atlas.cz</A>&gt;


    Já vidím tohle:


          rr = &quot;'\\xc3\\xbd \\xc3\\xb7 \\xc3\\xad \\xc3.'&quot;
          a = eval(rr)
          a

    '\xc3\xbd \xc3\xb7 \xc3\xad \xc3.'

          u = a.decode('utf-8')

    Traceback (most recent call last):
     File &quot;&lt;stdin&gt;&quot;, line 1, in &lt;module&gt;
     File &quot;c:\Python27\lib\encodings\utf_8.py&quot;, line 16, in decode
      return codecs.utf_8_decode(input, errors, True)
    UnicodeDecodeError: 'utf8' codec can't decode byte 0xc3 in position 9: 
invalid continuation byte


    To rr je okopírované tvoje repr.  Pomocí eval() to pøevedu na hodnotu a. 
Pak volám
    a.decode('utf-8') a ono to jinými slovy øíká, ¾e to není v utf-8.

    P.
    ______________________________________________________________

      Od: &quot;Jaroslav Lukesh&quot; &lt;<A HREF="http://www.py.cz/mailman/listinfo/python">lukesh na seznam.cz</A>&gt;
      Komu: Konference PyCZ &lt;<A HREF="http://www.py.cz/mailman/listinfo/python">python na py.cz</A>&gt;
      Datum: 02.05.2012 11:52
      Pøedmìt: Re: [python] Unicode/ne-unicode problem

      Je to promìnná, kde obsah byl získaný ze streamu dat.


            a='Ã½ Ã· Ã­ Ã.'
            repr(a)

      &quot;'\\xc3\\xbd \\xc3\\xb7 \\xc3\\xad \\xc3.'&quot;

      Dík, JL.


      ----- Pùvodní zpráva ----- Od: &quot;Petr Pøikryl&quot; &lt;<A HREF="http://www.py.cz/mailman/listinfo/python">prikryl na atlas.cz</A>&gt;


      Ten neèitelný øetìzec je zapsaný ve zdrojovém textu, nebo je naètený 
ze
      souboru?
      Kdy¾ mám v promìnné UTF-8 bajty, tak musím udìlat .decode() nebo 
unicode(),
      ale jako parametr musím uvést to 'utf-8'.

      Posloupnost bajtù v kódovaní UTF-8 je jednou z mo¾ných reprezentací 
Unicode
      øetìzce.  Vznikla tedy zakódováním (napøíklad u.encode('utf-8')) 
unicodového
      øetìzce.
      Kdy¾ z toho chci dostat zpìt Unicode øetìzec, musím provést 
dekódování,
      tak¾e u = s.decode('utf-8') nebo u = unicode(s, 'utf-8').

      Je mo¾né, ¾e je ta posloupnost bajtù poru¹ená, pak to mù¾e øvát, ¾e 
neví co
      s tím.
      Po¹li sem, jak vypadá repr(a) (pokud je to pro Python 2).

      Ptej se dál, ono se to vyjasní.

      P.

      ______________________________________________________________

        Od: &quot;Jaroslav Lukesh&quot; &lt;<A HREF="http://www.py.cz/mailman/listinfo/python">lukesh na seznam.cz</A>&gt;
        Komu: Konference PyCZ &lt;<A HREF="http://www.py.cz/mailman/listinfo/python">python na py.cz</A>&gt;
        Datum: 27.04.2012 10:35
        Pøedmìt: [python] Unicode/ne-unicode problem

        Dobrý den,

        potøeboval bych dostat z binárních dat utf8 na vstupu normální 
neunicodový
        výstup, ale nìjak mi to nejde.

        Vstup: znaky &quot;ì ¹ è ø&quot; v utf8 ji¾ v binární formì, systém je v 
iso8859-2

        Èekám ¾e z toho nìjak dostanu &quot;ì ¹ è ø&quot; ale poøád nic. Jak mu øíct, 
¾e 'Ã½
        Ã· Ã­ Ã.' je u¾ v utf8?

        Dìkuji, JL.


              a='Ã½ Ã· Ã­ Ã.'
              A=unicode(a,'iso8859-2')
              print a

        Ã½ Ã· Ã­ Ã.

              A=unicode(a,'utf8')

        Traceback (most recent call last):
         File &quot;&lt;stdin&gt;&quot;, line 1, in ?
         File &quot;/var/zope/python/lib/python2.4/encodings/utf_8.py&quot;, line 16, 
in
        decode
          return codecs.utf_8_decode(input, errors, True)
        UnicodeDecodeError: 'utf8' codec can't decode bytes in position 
9-10:
        invalid data


              A.encode('iso8859-2')

        '\xc3\xbd \xc3\xb7 \xc3\xad \xc3.'


              a.encode('iso8859-2')

        Traceback (most recent call last):
         File &quot;&lt;stdin&gt;&quot;, line 1, in ?
         File &quot;/var/zope/python/lib/python2.4/encodings/iso8859_2.py&quot;, line 
18, in
        encode
          return codecs.charmap_encode(input,errors,encoding_map)
        UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 
0:
        ordinal not in range(128)





              A.decode('iso8859-2')

        Traceback (most recent call last):
         File &quot;&lt;stdin&gt;&quot;, line 1, in ?
         File &quot;/var/zope/python/lib/python2.4/encodings/iso8859_2.py&quot;, line 
22, in
        decode
          return codecs.charmap_decode(input,errors,decoding_map)
        UnicodeEncodeError: 'ascii' codec can't encode characters in 
position 0-1:
        ordinal not in range(128)

              a.decode('iso8859-2')

        u'\u0102\u02dd \u0102\u02c7 \u0102\xad \u0102.'

              a.decode('utf8')

        Traceback (most recent call last):
         File &quot;&lt;stdin&gt;&quot;, line 1, in ?
         File &quot;/var/zope/python/lib/python2.4/encodings/utf_8.py&quot;, line 16, 
in
        decode
          return codecs.utf_8_decode(input, errors, True)
        UnicodeDecodeError: 'utf8' codec can't decode bytes in position 
9-10:
        invalid data





        _______________________________________________
        Python mailing list
        <A HREF="http://www.py.cz/mailman/listinfo/python">Python na py.cz</A>
        <A HREF="http://www.py.cz/mailman/listinfo/python">http://www.py.cz/mailman/listinfo/python</A>


      _______________________________________________
      Python mailing list
      <A HREF="http://www.py.cz/mailman/listinfo/python">Python na py.cz</A>
      <A HREF="http://www.py.cz/mailman/listinfo/python">http://www.py.cz/mailman/listinfo/python</A>

      _______________________________________________
      Python mailing list
      <A HREF="http://www.py.cz/mailman/listinfo/python">Python na py.cz</A>
      <A HREF="http://www.py.cz/mailman/listinfo/python">http://www.py.cz/mailman/listinfo/python</A>


    _______________________________________________
    Python mailing list
    <A HREF="http://www.py.cz/mailman/listinfo/python">Python na py.cz</A>
    <A HREF="http://www.py.cz/mailman/listinfo/python">http://www.py.cz/mailman/listinfo/python</A>
    _______________________________________________
    Python mailing list
    <A HREF="http://www.py.cz/mailman/listinfo/python">Python na py.cz</A>
    <A HREF="http://www.py.cz/mailman/listinfo/python">http://www.py.cz/mailman/listinfo/python</A>





------------------------------------------------------------------------------


  _______________________________________________
  Python mailing list
  <A HREF="http://www.py.cz/mailman/listinfo/python">Python na py.cz</A>
  <A HREF="http://www.py.cz/mailman/listinfo/python">http://www.py.cz/mailman/listinfo/python</A> 
------------- dal¹í èást ---------------
HTML pøíloha byla odstranìna...
URL: &lt;<A HREF="http://www.py.cz/pipermail/python/attachments/20120502/24bbf158/attachment.html">http://www.py.cz/pipermail/python/attachments/20120502/24bbf158/attachment.html</A>&gt;
</PRE>


<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Pøedchozí pøíspìvek: <A HREF="http://www.py.cz/pipermail/python/2012-May/010678.html">[python] Unicode/ne-unicode problem
</A></li>
	<LI>Následující pøíspìvek: <A HREF="http://www.py.cz/pipermail/python/2012-May/010680.html">[python] Unicode/ne-unicode problem
</A></li>
         <LI> <B>Zprávy tøídìné podle:</B> 
              <a href="http://www.py.cz/pipermail/python/2012-May/date.html#10679">[ data ]</a>
              <a href="http://www.py.cz/pipermail/python/2012-May/thread.html#10679">[ vlákna ]</a>
              <a href="http://www.py.cz/pipermail/python/2012-May/subject.html#10679">[ subjektu ]</a>
              <a href="http://www.py.cz/pipermail/python/2012-May/author.html#10679">[ autora ]</a>
         </LI>
       </UL>
<hr>
<a href="http://www.py.cz/mailman/listinfo/python">Dal¹í informace o konferenci Python</a><br>
</body>
<!-- Mirrored from www.py.cz/pipermail/python/2012-May/010679.html by HTTrack Website Copier/3.x [XR&CO'2014], Sun, 20 Mar 2016 21:17:31 GMT -->
</html>
